R/annotate_signature.R

```diff
@@ -1,61 +1,153 @@
-fetch_geneset <- function(type = "KEGG", max_size = Inf, min_size = 0){
+fetch_geneset <- function(type = "KEGG", username = NULL, access_date = NULL,
+                          max_size = Inf, min_size = 0){
 
-  # make sure type is one of "GO" or "KEGG"
-  if (!type %in% c("GO", "KEGG")) {
-    stop("type can only be either GO or KEGG.")
+  # make sure type is one of "GO", "KEGG", or "User"
+  if (!type %in% c("GO", "KEGG", "User")) {
+    stop("type can only be GO or KEGG or User.")
+  }
+
+  # username cannot be NULL if type is User
+  if (type == "User" & is.null(username)) {
+    stop("Please provide the TRIBE username.")
   }
 
   # to prevent overloading the TRIBE webserver, we limit the download size
-  # to 2000 gene sets at one time and make multiple downloads through changing
+  # to 200 gene sets at one time and make multiple downloads through changing
   # the offset until all gene sets have been downloaded.
-  request_limit <- 2000
+  request_limit <- 200
   request_time <- 1
   gene_sets_df_list <- list()
+
   while (TRUE) {
     # set request offset to the number of request limits that have already
     # been reached
     request_offset <- (request_time - 1) * request_limit
 
-    tribe_req <- httr::GET("http://tribe.greenelab.com/api/v1/geneset/",
-                           query = list(title__startswith = type,
-                                        show_tip = "true",
-                                        organism = "9",  # P.a. is number 9
-                                        xrdb = "Symbol",
-                                        limit = request_limit,
-                                        offset = request_offset,
-                                        format = "json"))
-
-    tribe_content <- suppressWarnings(httr::content(tribe_req, as = "text"))
-    gene_sets <- jsonlite::fromJSON(tribe_content)
-    gene_sets_df <- dplyr::data_frame(title = gene_sets$objects$title,
-                                      count = gene_sets$objects$tip_item_count,
-                                      genes = gene_sets$objects$tip$genes)
-    gene_sets_df_list[[request_time]] <- gene_sets_df
-
-    # the actual number of returned gene sets
-    returned_size <- nrow(gene_sets_df)
-    # If the return_size reaches the request limit, it indicates that we haven't
-    # downloaded all gene sets yet. We will make a new request repeatly until
-    # the number of returned gene sets is smaller than the request limit.
-    if (returned_size < request_limit) {
-      break
+    if (is.null(access_date)) {
+      # if access_date is NULL, use the tip version
+
+      if (type == "User") {
+        # filter gene set by creator's username
+        tribe_req <- httr::GET("https://tribe.greenelab.com/api/v1/geneset/",
+                               query = list(creator__username = username,
+                                            show_tip = "true",
+                                            organism = "9",  # P.a. is number 9
```

Comment:
Do you have to use an index (`organism = "9"`) here rather than the organism name? Just a consideration for the future if the order changes with the addition of more organisms.